Update SharedInfrastructure for Xunit V3

stefannikolei opened the pull request:
Submodule points to my fork for iterating faster. When this works, I will create a mr in the SharedInfrastructure project

stefannikolei committed:
Update SharedInfrastructure for Xunit V3

# Conflicts:
#	shared-infrastructure

stefannikolei commented:
I started looking into it. 

Gonna take longer. This is not a small change.

stefannikolei committed:
Migrate tests from xUnit v2 to xUnit.v3 3.2.2

Breaking changes addressed:
- Update xunit.v3 to 3.2.2 and xunit.runner.visualstudio to 3.1.5
- Remove Microsoft.DotNet.XUnitExtensions (conflicts with xunit.v3 via xunit.core v2)
- Replace ConditionalFact/ConditionalTheory with Fact/Theory SkipUnless/SkipWhen/SkipType
- Remove 'using Xunit.Abstractions' (ITestOutputHelper now in Xunit namespace globally)
- Fix TestFrameworkAttribute: string args → typeof(T) form
- Fix XunitTestFramework ctor: no longer takes IMessageSink
- Fix BeforeAfterTestAttribute Before/After: added IXunitTest parameter, use Xunit.v3 ns
- Fix DataAttribute: new GetData signature (DisposalTracker), add SupportsDiscoveryEnumeration()
- Fix IXunitSerializable/IXunitSerializationInfo: add 'using Xunit.Sdk' to all consumers
- Fix BasicSerializer: implement new IXunitSerializationInfo.GetValue(string) returning object?
  Store type info in dump format (key:TypeAQN:value) to enable typed reconstruction
- Fix TheoryData<T> CS0121 ambiguity on C# 12/net8.0: add explicit casts (T)value
  Affects ExifValuesTests, PixelOperationsTests, QuantizerTests, DitherTests,
  TestImageProviderTests, RotateTests, L8Tests, La16Tests
- Fix PngEncoderTests: TheoryData<T> now yields TheoryDataRow<T>, not object[]

stefannikolei committed:
Fix xunit.v3 test failures post-migration

stefannikolei committed:
Update shared-infra

stefannikolei committed:
Fix xunit skip

when used in nameof it is not allowed to be a filed it must be a property

JimBobSquarePants commented on tests/ImageSharp.Tests/Image/ImageTests.Decode_Cancellation.cs:
```diff
@@ -58,7 +58,7 @@ private static TheoryData<bool, string, double> CreateLoadData()
 
         // TODO: Figure out cancellation failures on Linux
-        [ConditionalTheory(typeof(TestEnvironment), nameof(TestEnvironment.IsWindows))]
+        [Theory(Skip = "Skipped on non-Windows platforms", SkipType = typeof(TestEnvironment), SkipUnless = nameof(TestEnvironment.IsWindows))]
```
`[ConditionalTheory]` comes from Arcade doesn't it. Can we write our own version? Seems less verbose.

stefannikolei replied:
I can look into it. Arcade is strange. You use a package which is not findable on Nuget

JimBobSquarePants replied:
Yeah it’s from a runtime feed we add via the submodule. Has RemoteExecutor also.

stefannikolei replied:
I can add it. Where would you want me to put it? TestUtilities.XUnit.Attributes ?

JimBobSquarePants commented on tests/ImageSharp.Tests/Memory/Allocators/UniformUnmanagedMemoryPoolTests.Trim.cs:
```diff
@@ -46,18 +48,16 @@ static void RunTest()
             // TODO: Investigate failures on macOS. All handles are released after GC.
             // (It seems to happen more consistently on .NET 6.)
-            [ConditionalFact(nameof(IsNotMacOS))]
+            [Fact(Skip = "Skipped on macOS", SkipUnless = nameof(IsNotMacOS))]
```
Same here, can we roll our own?

JimBobSquarePants commented on tests/ImageSharp.Tests/PixelFormats/PixelOperations/PixelOperationsTests.cs:
```diff
@@ -315,35 +314,35 @@ public void ToVector4(int count)
-        new TestPixel<RgbaVector>(),
-        new TestPixel<Short2>(),
-        new TestPixel<Short4>(),
+        (object)new TestPixel<A8>(),
```
Do we have to explicitly cast now?

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/BasicSerializer.cs:
```diff
@@ -41,8 +44,18 @@ private Type LoadDump(string dump)
         {
-            string[] kv = s.Split(Separator);
-            this.map[kv[0]] = kv[1];
+            // Format: key:TypeAQN:value  (3 parts max to preserve ':' inside value)
```
I'll need you to explain all this I'm afraid.

stefannikolei replied:
The old interface did  pass the type in the Get. That's why it was saved internally as key:value.

With XunitV3 we do not get that info with the get method that's why we now save it in the dictionary. So we now have 3 parts and not 3 parts like before

stefannikolei committed:
Remove legacy way in BasicSerializer

stefannikolei committed:
Remove cast to object

stefannikolei committed:
Port ConditionalTheory and Fact

Source of it: https://github.com/dotnet/arcade/tree/main/src/Microsoft.DotNet.XUnitExtensions.Shared/Attributes

JimBobSquarePants committed:
Merge branch 'main' into sn/upd/xunit

JimBobSquarePants commented on tests/ImageSharp.Tests/Memory/Allocators/UniformUnmanagedMemoryPoolTests.Trim.cs:
```diff
@@ -96,21 +102,25 @@ static void RunTest()
         }
 
-        public static readonly bool Is32BitProcess = !Environment.Is64BitProcess;
+        public static bool Is32BitProcess => !Environment.Is64BitProcess;
```
I've spotted the removal of `readonly` in a few places in the diff. why is that?

JimBobSquarePants commented on tests/ImageSharp.Tests/Memory/Allocators/UniformUnmanagedMemoryPoolTests.Trim.cs:
```diff
@@ -129,11 +139,16 @@ static void RunTest()
                 GCMemoryInfo memInfo = GC.GetGCMemoryInfo();
                 int highLoadThreshold = (int)(memInfo.HighMemoryLoadThresholdBytes / oneMb);
-                highLoadThreshold = (int)(trimSettings.HighPressureThresholdRate * highLoadThreshold);
+                highLoadThreshold = (int)(
```
Formatting like this is weird. Can you please look in the diff for other odd changes like this and revert. I've spotted a few.

JimBobSquarePants commented on tests/ImageSharp.Tests/PixelFormats/L8Tests.cs:
```diff
@@ -12,7 +12,7 @@ namespace SixLabors.ImageSharp.Tests.PixelFormats;
 {
     public static readonly TheoryData<byte> LuminanceData
-        = new() { 0, 1, 2, 3, 5, 13, 31, 71, 73, 79, 83, 109, 127, 128, 131, 199, 250, 251, 254, 255 };
+        = new() { (byte)0, (byte)1, (byte)2, (byte)3, (byte)5, (byte)13, (byte)31, (byte)71, (byte)73, (byte)79, (byte)83, (byte)109, (byte)127, (byte)128, (byte)131, (byte)199, (byte)250, (byte)251, (byte)254, (byte)255 };
```
And we definitely need this new casting?

stefannikolei replied:
<img width="643" height="170" alt="image" src="https://github.com/user-attachments/assets/b8a1c189-af07-4fb4-8631-c5e90a08d25c" />

Do you see another way?

stefannikolei replied:
<img width="367" height="59" alt="image" src="https://github.com/user-attachments/assets/89316ed8-da55-42c3-a350-2cf1aedb1892" />
better than this

JimBobSquarePants replied:
Does casting the first entry work? `(byte)0`

stefannikolei replied:
Like this? 
<img width="543" height="79" alt="image" src="https://github.com/user-attachments/assets/36e4cb2e-5a58-42d3-82c8-1eebf8dc7856" />
 nope ...

JimBobSquarePants replied:
🤮

Oh XUnit!

JimBobSquarePants commented on tests/ImageSharp.Tests/Processing/Processors/Quantization/QuantizerTests.cs:
```diff
@@ -74,79 +74,79 @@ public static readonly TheoryData<IQuantizer> Quantizers
-        new WebSafePaletteQuantizer(OrderedDitherOptions),
-        new WernerPaletteQuantizer(OrderedDitherOptions),
-        new WuQuantizer(OrderedDitherOptions)
+        (IQuantizer)KnownQuantizers.Hexadecatree,
```
Same here. Lots of new casting

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/Attributes/ImageDataAttributeBase.cs:
```diff
@@ -59,11 +71,14 @@ public override IEnumerable<object[]> GetData(MethodInfo testMethod)
                     {
-                        addedRows = memberItems.Select(x => new[] { x });
-                    }
+                        if (x is ITheoryDataRow row) return row.GetData();
```
braces

stefannikolei replied:
There should be an editorconfig setting for this. 

Another thing is formatting. 

Should we work onto an ci check and tooling that formats the code? Or do you want it handcrafted. 

I made good experience with csharpier. It's opinionated. At first a bit strange but after getting used to it 👌 and no more formatting discussions or problems

JimBobSquarePants replied:
>Should we work onto an ci check and tooling that formats the code? Or do you want it handcrafted.

No. I do not want to ever rely on automation for code formatting. I want people to be aware of the styles and standards. 

>There should be an editorconfig setting for this.

`csharp_prefer_braces` is already present as a warning. We just don't have them as errors in the tests.

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/Tests/TestEnvironmentTests.cs:
```diff
@@ -28,21 +25,24 @@ private void CheckPath(string path)
     [Fact]
-    public void SolutionDirectoryFullPath()
-        => this.CheckPath(TestEnvironment.SolutionDirectoryFullPath);
+    public void SolutionDirectoryFullPath() =>
```
Odd formatting changes

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/Tests/TestEnvironmentTests.cs:
```diff
@@ -28,21 +25,24 @@ private void CheckPath(string path)
     public void GetReferenceOutputFileName()
     {
-        string actual = Path.Combine(TestEnvironment.ActualOutputDirectoryFullPath, @"foo\bar\lol.jpeg");
+        string actual = Path.Combine(
```
No need to wrap here.

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/Tests/TestEnvironmentTests.cs:
```diff
@@ -72,7 +75,10 @@ public void GetReferenceEncoder_ReturnsCorrectEncoders_Windows(string fileName,
     [InlineData("lol/Baz.gif", typeof(GifDecoder))]
     [InlineData("lol/foobar.webp", typeof(WebpDecoder))]
-    public void GetReferenceDecoder_ReturnsCorrectDecoders_Windows(string fileName, Type expectedDecoderType)
+    public void GetReferenceDecoder_ReturnsCorrectDecoders_Windows(
```
No here etc.

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/XUnit/ConditionalFact.cs:
```diff
@@ -0,0 +1,38 @@
+            CalleeType = calleeType;
+            ConditionMemberNames = conditionMemberNames;
+            string skipReason = ConditionalTestDiscoverer.EvaluateSkipConditions(calleeType, conditionMemberNames);
+            if (skipReason != null)
```
braces

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/XUnit/ConditionalFact.cs:
```diff
@@ -0,0 +1,38 @@
+                Skip = skipReason;
+        }
+
+        [Obsolete(
```
odd wrapping

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/XUnit/ConditionalTestDiscoverer.cs:
```diff
@@ -0,0 +1,146 @@
+            List<string> falseConditions = new(conditionMemberNames.Length);
+            foreach (string entry in conditionMemberNames)
+            {
+                if (string.IsNullOrWhiteSpace(entry))
```
braces

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/XUnit/ConditionalTheory.cs:
```diff
@@ -0,0 +1,42 @@
+                conditionMemberNames
+            );
+            if (skipReason != null)
+                Skip = skipReason;
```
braces

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/XUnit/ConditionalTheory.cs:
```diff
@@ -0,0 +1,42 @@
+                Skip = skipReason;
+        }
+
+        [Obsolete(
```
odd wrapping

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/XUnit/StaticReflectionConstants.cs:
```diff
@@ -0,0 +1,64 @@
+        public DynamicallyAccessedMemberTypes MemberTypes { get; }
+    }
+
+    internal enum DynamicallyAccessedMemberTypes
```
This should be available to you without copying.

https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.dynamicallyaccessedmembertypes?view=net-8.0

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/TestEnvironment.cs:
```diff
@@ -68,12 +73,15 @@ private static string GetSolutionDirectoryFullPathImpl()
             if (directory == null)
             {
-                throw new DirectoryNotFoundException($"Unable to find ImageSharp solution directory from {TestAssemblyFile}!");
+                throw new DirectoryNotFoundException(
```
unnecessary wrapping

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/TestEnvironment.cs:
```diff
@@ -68,12 +73,15 @@ private static string GetSolutionDirectoryFullPathImpl()
                     $"Unable to find ImageSharp solution directory from {TestAssemblyFile} because of {ex.GetType().Name}!",
-                    ex);
+                    ex
+                );
```
I've seen this change in a few places now and it violates StyleCop

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/TestEnvironment.cs:
```diff
@@ -92,25 +100,29 @@ private static string GetFullPath(string relativePath) =>
     internal static bool IsMono => Type.GetType("Mono.Runtime") != null; // https://stackoverflow.com/a/721194
 
-    internal static bool IsWindows => RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+    public static bool IsWindows => RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
```
Why is this public and no others?

stefannikolei replied:
This was needed for using plain fact. This can probably switched back to internal

JimBobSquarePants commented on tests/ImageSharp.Tests/TestUtilities/TestEnvironment.cs:
```diff
@@ -164,8 +174,10 @@ private static void PrepareRemoteExecutor()
 
-        string remoteExecutorConfigPath =
-            Path.Combine(TestAssemblyFile.DirectoryName, "Microsoft.DotNet.RemoteExecutor.exe.config");
+        string remoteExecutorConfigPath = Path.Combine(
```
More new wrapping

JimBobSquarePants commented:
@stefannikolei How can I help here?

stefannikolei commented:
> @stefannikolei How can I help here? 

I'll look into it today or tomorrow. You could update shared Infra to the latest xunitv3 v3 version (no v4 please there are braking changes 😜)
